modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -0,0 +1,718 @@
+
+        cacheOrg.put(101, new Organization("o1"));
+        cacheOrg.put(102, new Organization("o2"));
+        cachePers.put(103, new Person(101, "p1"));
+        cachePers.put(104, new Person(102, "p2"));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /** Checks successful JDBC queries. */
+    @Test
+    public void testJdbcQuery() throws Exception {
+        for (int i = 0; i < 2; i++) {
+            jdbcQuery(SQL);
+
+            checkSqlPlanHistory(1);
+        }
+    }
+
+    /** Checks failed JDBC queries. */
+    @Test
+    public void testJdbcQueryFailed() throws Exception {
+        try {
+            jdbcQuery(SQL_FAILED);
+        }
+        catch (Exception e) {
+            if (e instanceof AssumptionViolatedException)
+                throw e;
+        }
+
+        checkSqlPlanHistory(1);
+    }
+
+    /** Checks successful SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsQuery() {
+        runSuccessfulQuery(sqlFieldsQry);
+    }
+
+    /** Checks failed SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsQueryFailed() {
+        runFailedQuery(sqlFieldsQryFailed);
+    }
+
+    /** Checks successful cross-cache SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsCrossCacheQuery() {
+        runSuccessfulQuery(sqlFieldsQryCrossCache);
+    }
+
+    /** Checks failed cross-cache SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsCrossCacheQueryFailed() {
+        runFailedQuery(sqlFieldsQryCrossCacheFailed);
+    }
+
+    /** Checks successful SqlFieldsQuery with reduce phase. */
+    @Test
+    public void testSqlFieldsQueryWithReducePhase() {
+        runQueryWithReducePhase(() -> {
+            cacheQuery(sqlFieldsQryWithReducePhase.setLocal(loc), "pers");
+
+            checkSqlPlanHistory((!isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) ? 3 : 1);
+        });
+    }
+
+    /** Checks failed SqlFieldsQuery with reduce phase. */
+    @Test
+    public void testSqlFieldsQueryWithReducePhaseFailed() {
+        runQueryWithReducePhase(() -> {
+            try {
+                cacheQuery(sqlFieldsQryWithReducePhaseFailed.setLocal(loc), "pers");
+            }
+            catch (Exception ignore) {
+                //No-Op
+            }
+
+            checkSqlPlanHistory((!isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) ? 1 : 0);
+        });
+    }
+
+    /** Checks successful SqlQuery. */
+    @Test
+    public void testSqlQuery() {
+        runSuccessfulQuery(sqlQry);
+    }
+
+    /** Checks failed SqlQuery. */
+    @Test
+    public void testSqlQueryFailed() {
+        runFailedQuery(sqlQryFailed);
+    }
+
+    /** Checks ScanQuery. */
+    @Test
+    public void testScanQuery() {
+        runQueryWithoutPlan(scanQry);
+    }
+
+    /** Checks TextQuery. */
+    @Test
+    public void testTextQuery() {
+        runQueryWithoutPlan(textQry);
```

Comment:
Apparently, for caches with built-in classes as data types, marking those classes as indexed types in the cache configuration is sufficient for text queries to work.

However, this is insufficient when custom classes are used as cache data types. In such cases, `@QueryTextField` annotations for the corresponding custom class fields will indeed be required.

There is a check in the `cacheQuery()` method that ensures all queries return a result, including those executed without a plan. If necessary, I can rewrite the text query test so that it uses a query from the `Person` cache and therefore a `@QueryTextField` annotation will be needed.